processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+  private static final float MAX_LOAD_FACTOR = 0.75f;
+  private static final int HASH_SIZE = Ints.BYTES;
+
+  private final ByteBuffer buffer;
+  private final KeySerde<KeyType> keySerde;
+  private final int keySize;
+  private final BufferAggregator[] aggregators;
+  private final int[] aggregatorOffsets;
+  private final int bucketSize;
+  private final int tableArenaSize;
+  private final int maxBufferGrouperSize; // Integer.MAX_VALUE in production, only used for unit tests
+
+  // Buffer pointing to the current table (it moves around as the table grows)
+  private ByteBuffer tableBuffer;
+
+  // Offset of tableBuffer within the larger buffer
+  private int tableStart;
+
+  // Current number of buckets in the table
+  private int buckets;
+
+  // Number of elements in the table right now
+  private int size;
+
+  // Maximum number of elements in the table before it must be resized
+  private int maxSize;
+
+  public BufferGrouper(
+      final ByteBuffer buffer,
+      final KeySerde<KeyType> keySerde,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final int maxBufferGrouperSize
+  )
+  {
+    this.buffer = buffer;
+    this.keySerde = keySerde;
+    this.keySize = keySerde.keySize();
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+    this.maxBufferGrouperSize = maxBufferGrouperSize;
+
+    int offset = HASH_SIZE + keySize;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+
+    this.bucketSize = offset;
+    this.tableArenaSize = buffer.capacity() / (bucketSize + Ints.BYTES) * bucketSize;
+
+    reset();
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+    if (keyBuffer == null) {
+      return false;
+    }
+
+    Preconditions.checkArgument(
+        keyBuffer.remaining() == keySize,
+        "key size[%s] != keySize[%s]",
+        keyBuffer.remaining(),
+        keySize
+    );
+
+    int bucket = findBucket(
+        tableBuffer,
+        buckets,
+        bucketSize,
+        size < Math.min(maxSize, maxBufferGrouperSize),
+        keyBuffer,
+        keySize,
+        keyHash
+    );
+
+    if (bucket < 0) {
+      if (size < maxBufferGrouperSize) {
+        growIfPossible();
+        bucket = findBucket(tableBuffer, buckets, bucketSize, size < maxSize, keyBuffer, keySize, keyHash);
+      }
+
+      if (bucket < 0) {
+        return false;
+      }
+    }
+
+    final int offset = bucket * bucketSize;
+
+    // Set up key if this is a new bucket.
+    if (!isUsed(bucket)) {
+      tableBuffer.position(offset);
+      tableBuffer.putInt(keyHash | 0x80000000);
```

Comment:
This is because it's not always possible to re-compute the hash when growing the hash table (it's based on the hashCode of an actual Java object, which accessible on "put" but is not accessible on resize)